File: agent/packages/schema/src/forge.ts
/**
 * 炼器（武器）IPC 共享原子（plan-forge-v1 §4 数据契约）。
 *
 * 服务端 → 客户端的 forge_* 推送（砧/会话/结算/图谱书）走 server-data.ts；
 * 客户端 → 服务端的 forge_* 操作（起炉/击键/铭文/开光/翻页/学图谱/放砧）走 client-request.ts。
 * 本文件提供二者共享的原子类型。
 */
import { Type, type Static } from "@sinclair/typebox";

import { ColorKind } from "./cultivation.js";

/** plan §1.3 四步串行（与服务端 ForgeStep 对齐）。 */
export const ForgeStep = Type.Union(
  [
    Type.Literal("billet"),
    Type.Literal("tempering"),
    Type.Literal("inscription"),
    Type.Literal("consecration"),
    Type.Literal("done"),
  ],
  { description: "plan §1.3 四步进程" },
);
export type ForgeStep = Static<typeof ForgeStep>;

/** plan §2 品阶四阶。 */
export const WeaponTier = Type.Union(
  [
    Type.Literal(1),
    Type.Literal(2),
    Type.Literal(3),
    Type.Literal(4),
  ],
  { description: "plan §2 品阶：1=凡器/2=法器/3=灵器/4=道器" },
);
export type WeaponTier = Static<typeof WeaponTier>;

/** 淬炼击键（J=Light, K=Heavy, L=Fold）。 */
export const TemperBeat = Type.Union(
  [
    Type.Literal("L"),
    Type.Literal("H"),
    Type.Literal("F"),
  ],
  { description: "plan §1.3.2 淬炼击键" },
);
export type TemperBeat = Static<typeof TemperBeat>;

/** plan §1.3 五结果桶。 */
export const ForgeOutcomeBucket = Type.Union(
  [
    Type.Literal("perfect"),
    Type.Literal("good"),
    Type.Literal("flawed"),
    Type.Literal("waste"),
    Type.Literal("explode"),
  ],
  { description: "plan §1.3 结果分桶" },
);
export type ForgeOutcomeBucket = Static<typeof ForgeOutcomeBucket>;

/** 坯料步实时状态。 */
export const ForgeStepBilletState = Type.Object(
  {
    step: Type.Literal("billet"),
    materials_in: Type.Array(
      Type.Tuple([Type.String(), Type.Integer({ minimum: 0 })]),
    ),
    active_carrier: Type.Optional(Type.String()),
    resolved_tier_cap: Type.Integer({ minimum: 1 }),
  },
  { additionalProperties: false },
);

/** 淬炼步实时状态。 */
export const ForgeStepTemperingState = Type.Object(
  {
    step: Type.Literal("tempering"),
    pattern: Type.Array(TemperBeat),
    beat_cursor: Type.Integer({ minimum: 0 }),
    hits: Type.Integer({ minimum: 0 }),
    misses: Type.Integer({ minimum: 0 }),
    deviation: Type.Integer({ minimum: 0 }),
    qi_spent: Type.Number({ minimum: 0 }),
  },
  { additionalProperties: false },
);

/** 铭文步实时状态。 */
export const ForgeStepInscriptionState = Type.Object(
  {
    step: Type.Literal("inscription"),
    filled_slots: Type.Integer({ minimum: 0 }),
    max_slots: Type.Integer({ minimum: 0 }),
    failed: Type.Boolean(),
  },
  { additionalProperties: false },
);

/** 开光步实时状态。 */
export const ForgeStepConsecrationState = Type.Object(
  {
    step: Type.Literal("consecration"),
    qi_injected: Type.Number({ minimum: 0 }),
    qi_required: Type.Number({ minimum: 0 }),
    color_imprint: Type.Optional(ColorKind),
  },
  { additionalProperties: false },
);

/** 无步骤状态。 */
export const ForgeStepNoneState = Type.Object(
  {
    step: Type.Literal("none"),
  },
  { additionalProperties: false },
);

/** 各步实时状态（tagged union）。 */
export const ForgeStepState = Type.Union(
  [
    ForgeStepBilletState,
    ForgeStepTemperingState,
    ForgeStepInscriptionState,
    ForgeStepConsecrationState,
    ForgeStepNoneState,
  ],
  { description: "plan §1.3 各步实时状态" },
);
export type ForgeStepState = Static<typeof ForgeStepState>;

/** 单条已学图谱条目。 */
export const ForgeBlueprintEntryV1 = Type.Object(
  {
    id: Type.String(),
    display_name: Type.String(),
    tier_cap: Type.Integer({ minimum: 1, maximum: 4 }),
    step_count: Type.Integer({ minimum: 0 }),
  },
  { additionalProperties: false },
);
export type ForgeBlueprintEntryV1 = Static<typeof ForgeBlueprintEntryV1>;

// ─── server → client payload 结构体（server-data.ts 中组装） ──────────

/** 砧信息快照。 */
export const WeaponForgeStationDataV1 = Type.Object(
  {
    station_id: Type.String(),
    tier: Type.Integer({ minimum: 1, maximum: 4 }),
    integrity: Type.Number({ minimum: 0, maximum: 1 }),
    owner_name: Type.String(),
    has_session: Type.Boolean(),
  },
  { additionalProperties: false },
);
export type WeaponForgeStationDataV1 = Static<typeof WeaponForgeStationDataV1>;

/** 锻造会话实时状态。 */
export const ForgeSessionDataV1 = Type.Object(
  {
    session_id: Type.Integer({ minimum: 0 }),
    blueprint_id: Type.String(),
    blueprint_name: Type.String(),
    active: Type.Boolean(),
    current_step: ForgeStep,
    step_index: Type.Integer({ minimum: 0 }),
    achieved_tier: Type.Integer({ minimum: 0, maximum: 4 }),
    step_state: ForgeStepState,
  },
  { additionalProperties: false },
);
export type ForgeSessionDataV1 = Static<typeof ForgeSessionDataV1>;

/** 锻造结果结算推送。 */
export const ForgeOutcomeDataV1 = Type.Object(
  {
    session_id: Type.Integer({ minimum: 0 }),
    blueprint_id: Type.String(),
    bucket: ForgeOutcomeBucket,
    weapon_item: Type.Optional(Type.String()),
    quality: Type.Number({ minimum: 0, maximum: 1 }),
    color: Type.Optional(ColorKind),
    side_effects: Type.Array(Type.String()),
    achieved_tier: Type.Integer({ minimum: 0, maximum: 4 }),
    flawed_path: Type.Boolean(),
  },
  { additionalProperties: false },
);
export type ForgeOutcomeDataV1 = Static<typeof ForgeOutcomeDataV1>;

/** 已学图谱书快照。 */
export const ForgeBlueprintBookDataV1 = Type.Object(
  {
    learned: Type.Array(ForgeBlueprintEntryV1),
    current_index: Type.Integer({ minimum: 0 }),
  },
  { additionalProperties: false },
);
export type ForgeBlueprintBookDataV1 = Static<typeof ForgeBlueprintBookDataV1>;

File: server/src/schema/forge.rs
//! 炼器（武器）IPC 共享原子（plan-forge-v1 §4 数据契约）。
//!
//! 与 TypeScript `agent/packages/schema/src/forge.ts` 1:1 镜像。
//! `server-data` / `client-request` 模块复用本文件的类型组装具体 payload。

use serde::{Deserialize, Serialize};

use crate::cultivation::components::ColorKind;

/// plan §1.3 四步串行（与 `ForgeSession::current_step` 对齐）。
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum ForgeStepV1 {
    Billet,
    Tempering,
    Inscription,
    Consecration,
    Done,
}

/// plan §2 品阶四阶（与 `compute_achieved_tier` 输出对齐）。
#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum WeaponTierV1 {
    Mundane = 1,
    Artifact = 2,
    Spirit = 3,
    Dao = 4,
}

/// 淬炼击键（J=Light, K=Heavy, L=Fold）。
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
pub enum TemperBeatV1 {
    #[serde(rename = "L")]
    Light,
    #[serde(rename = "H")]
    Heavy,
    #[serde(rename = "F")]
    Fold,
}

impl From<crate::forge::blueprint::TemperBeat> for TemperBeatV1 {
    fn from(b: crate::forge::blueprint::TemperBeat) -> Self {
        match b {
            crate::forge::blueprint::TemperBeat::Light => Self::Light,
            crate::forge::blueprint::TemperBeat::Heavy => Self::Heavy,
            crate::forge::blueprint::TemperBeat::Fold => Self::Fold,
        }
    }
}

/// plan §1.3 五结果桶。
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum ForgeOutcomeBucketV1 {
    Perfect,
    Good,
    Flawed,
    Waste,
    Explode,
}

impl From<crate::forge::events::ForgeBucket> for ForgeOutcomeBucketV1 {
    fn from(b: crate::forge::events::ForgeBucket) -> Self {
        match b {
            crate::forge::events::ForgeBucket::Perfect => Self::Perfect,
            crate::forge::events::ForgeBucket::Good => Self::Good,
            crate::forge::events::ForgeBucket::Flawed => Self::Flawed,
            crate::forge::events::ForgeBucket::Waste => Self::Waste,
            crate::forge::events::ForgeBucket::Explode => Self::Explode,
        }
    }
}

// ─── server → client 推送 payload（plan §4 数据契约） ───────────────────

/// 砧信息快照。
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct WeaponForgeStationDataV1 {
    pub station_id: String,
    pub tier: u8,
    pub integrity: f32,
    pub owner_name: String,
    pub has_session: bool,
}

/// 锻造会话实时状态。
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct ForgeSessionDataV1 {
    pub session_id: u64,
    pub blueprint_id: String,
    pub blueprint_name: String,
    pub active: bool,
    pub current_step: ForgeStepV1,
    pub step_index: u32,
    pub achieved_tier: u32,
    pub step_state: ForgeStepStateDataV1,
}

/// 各步实时状态（按 current_step 选择对应 variant）。
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(tag = "step", rename_all = "snake_case")]
pub enum ForgeStepStateDataV1 {
    #[serde(rename = "billet")]
    Billet {
        /// 当前投入材料：{material -> count}
        materials_in: Vec<(String, u32)>,
        active_carrier: Option<String>,
        resolved_tier_cap: u32,
    },
    #[serde(rename = "tempering")]
    Tempering {
        /// 待击 pattern 序列
        pattern: Vec<TemperBeatV1>,
        beat_cursor: u32,
        hits: u32,
        misses: u32,
        deviation: u32,
        qi_spent: f64,
    },
    #[serde(rename = "inscription")]
    Inscription {
        filled_slots: u32,
        max_slots: u32,
        failed: bool,
    },
    #[serde(rename = "consecration")]
    Consecration {
        qi_injected: f64,
        qi_required: f64,
        color_imprint: Option<ColorKind>,
    },
    None,
}

/// 锻造结果结算推送。
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct ForgeOutcomeDataV1 {
    pub session_id: u64,
    pub blueprint_id: String,
    pub bucket: ForgeOutcomeBucketV1,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub weapon_item: Option<String>,
    pub quality: f32,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub color: Option<ColorKind>,
    pub side_effects: Vec<String>,
    pub achieved_tier: u32,
    /// 是否走了残缺匹配路径
    pub flawed_path: bool,
}

/// 已学图谱书快照。
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct ForgeBlueprintBookDataV1 {
    pub learned: Vec<ForgeBlueprintEntryV1>,
    pub current_index: u32,
}

/// 单条已学图谱条目（与 client BlueprintScrollStore 对齐）。
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct ForgeBlueprintEntryV1 {
    pub id: String,
    pub display_name: String,
    pub tier_cap: u8,
    pub step_count: u32,
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn forge_step_serde_snake_case() {
        assert_eq!(
            serde_json::to_string(&ForgeStepV1::Billet).unwrap(),
            "\"billet\""
        );
        assert_eq!(
            serde_json::to_string(&ForgeStepV1::Tempering).unwrap(),
            "\"tempering\""
        );
    }

    #[test]
    fn outcome_bucket_serde_roundtrip() {
        let s = serde_json::to_string(&ForgeOutcomeBucketV1::Explode).unwrap();
        assert_eq!(s, "\"explode\"");
        let back: ForgeOutcomeBucketV1 = serde_json::from_str("\"flawed\"").unwrap();
        assert_eq!(back, ForgeOutcomeBucketV1::Flawed);
    }

    #[test]
    fn forge_step_state_billet_roundtrip() {
        let state = ForgeStepStateDataV1::Billet {
            materials_in: vec![("iron_ingot".into(), 3)],
            active_carrier: None,
            resolved_tier_cap: 1,
        };
        let s = serde_json::to_string(&state).unwrap();
        let back: ForgeStepStateDataV1 = serde_json::from_str(&s).unwrap();
        assert_eq!(back, state);
    }

    #[test]
    fn forge_step_state_tempering_roundtrip() {
        let state = ForgeStepStateDataV1::Tempering {
            pattern: vec![TemperBeatV1::Light, TemperBeatV1::Heavy],
            beat_cursor: 0,
            hits: 0,
            misses: 0,
            deviation: 0,
            qi_spent: 0.0,
        };
        let s = serde_json::to_string(&state).unwrap();
        let back: ForgeStepStateDataV1 = serde_json::from_str(&s).unwrap();
        assert_eq!(back, state);
    }

    #[test]
    fn forge_outcome_data_roundtrip() {
        let outcome = ForgeOutcomeDataV1 {
            session_id: 1,
            blueprint_id: "iron_sword_v0".into(),
            bucket: ForgeOutcomeBucketV1::Perfect,
            weapon_item: Some("iron_sword".into()),
            quality: 1.0,
            color: None,
            side_effects: vec![],
            achieved_tier: 1,
            flawed_path: false,
        };
        let s = serde_json::to_string(&outcome).unwrap();
        let back: ForgeOutcomeDataV1 = serde_json::from_str(&s).unwrap();
        assert_eq!(back, outcome);
    }

    #[test]
    fn forge_station_data_roundtrip() {
        let data = WeaponForgeStationDataV1 {
            station_id: "s1".into(),
            tier: 1,
            integrity: 0.95,
            owner_name: "test".into(),
            has_session: false,
        };
        let s = serde_json::to_string(&data).unwrap();
        let back: WeaponForgeStationDataV1 = serde_json::from_str(&s).unwrap();
        assert_eq!(back, data);
    }

    #[test]
    fn temper_beat_serde() {
        assert_eq!(
            serde_json::to_string(&TemperBeatV1::Light).unwrap(),
            "\"L\""
        );
        assert_eq!(
            serde_json::to_string(&TemperBeatV1::Heavy).unwrap(),
            "\"H\""
        );
        assert_eq!(
            serde_json::to_string(&TemperBeatV1::Fold).unwrap(),
            "\"F\""
        );
        let back: TemperBeatV1 = serde_json::from_str("\"L\"").unwrap();
        assert_eq!(back, TemperBeatV1::Light);
    }

    #[test]
    fn blueprint_entry_roundtrip() {
        let entry = ForgeBlueprintEntryV1 {
            id: "iron_sword_v0".into(),
            display_name: "铁剑（测试）".into(),
            tier_cap: 1,
            step_count: 1,
        };
        let s = serde_json::to_string(&entry).unwrap();
        let back: ForgeBlueprintEntryV1 = serde_json::from_str(&s).unwrap();
        assert_eq!(back, entry);
    }
}
